execute/profiler_test.go

```diff
@@ -18,50 +18,136 @@ import (
-	// Need to have the goroutines to sync on their writes to the buffer.
-	var mu sync.Mutex
 	wantStr.WriteString(`
-#datatype,string,long,string,dateTime:RFC3339,dateTime:RFC3339,string,string,long
-#group,false,false,true,false,false,false,false,false
-#default,_profiler,,,,,,,
-,result,table,_measurement,Begin,End,Type,Label,Duration
+#datatype,string,long,string,string,string,long,long,long,long,double
+#group,false,false,true,false,false,false,false,false,false,false
+#default,_profiler,,,,,,,,,
+,result,table,_measurement,Type,Label,Count,MinDuration,MaxDuration,DurationSum,MeanDuration
 `)
-	// Unfortunately, the operator profiler result is only grouped on _measurement, it cannot
-	// ensure a deterministic row order with our executetest.EqualResultIterators
-	// Therefore, currently I only set this goroutine count to 2.
-	count := 2
+	count := 4
 	wg := sync.WaitGroup{}
 	wg.Add(count)
-	fn := func(label string, ctx context.Context) {
-		_, span := execute.StartSpanFromContext(ctx, "tf", label)
+	fn := func(opType string, label string, ctx context.Context, offset int) {
+		st := time.Date(2020, 10, 14, 12, 30, 0, 0, time.UTC)
+		_, span := execute.StartSpanFromContext(ctx, opType, label, opentracing.StartTime(st))
 		profilerSpan := span.(*execute.OperatorProfilingSpan)
 		// Finish() will write the data to the profiler
 		// In Flux runtime, this is called when an execution node finishes execution
-		profilerSpan.Finish()
-		mu.Lock()
-		// Write the expected result from raw span data.
-		wantStr.WriteString(fmt.Sprintf(",,0,profiler/operator,%s,%s,tf,%s,%d\n",
-			profilerSpan.Result.Start.Format(time.RFC3339Nano),
-			profilerSpan.Result.Stop.Format(time.RFC3339Nano),
-			profilerSpan.Result.Label,
-			profilerSpan.Result.Stop.Sub(profilerSpan.Result.Start).Nanoseconds()))
-		mu.Unlock()
+		profilerSpan.FinishWithOptions(opentracing.FinishOptions{
+			FinishTime: time.Date(2020, 10, 14, 12, 30, 0, 1000+offset, time.UTC),
+		})
 		wg.Done()
 	}
 	for i := 0; i < count; i++ {
-		go fn(fmt.Sprintf("op%d", i), ctx)
+		op := fmt.Sprintf("type%d", i%2)
+		go fn(op, "lab0", ctx, 100+i*i)
+		// Waiting between each loop seems to ensure better consistency of row order
+		// for final table. However, because the result aggregates are stored in a map,
+		// the order in which the rows are written cannot be guaranteed
+		time.Sleep(100 * time.Millisecond)
 	}
+	wantStr.WriteString(fmt.Sprintf(",,0,profiler/operator,%s,%s,%d,%d,%d,%d,%f\n",
+		"type0",
+		"lab0",
+		2,
+		1100,
+		1104,
+		2204,
+		1102.0,
+	))
+	wantStr.WriteString(fmt.Sprintf(",,0,profiler/operator,%s,%s,%d,%d,%d,%d,%f\n",
+		"type1",
+		"lab0",
+		2,
+		1101,
+		1109,
+		2210,
+		1105.0,
+	))
+	wg.Wait()
+	// Wait a bit for the profiling results to be added.
+	// In the query code path this is guaranteed because we only access the result
+	// after the query finishes execution AND its result tables are read and encoded.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Is there anything we can do to remove this `Sleep`? Seems like it is ripe for flakiness that causes the test to get skipped sometime in the future.

Reply:
I'm not sure why that's in there either. @yzhang1991 might have a better idea.

Reply:
Hmm... I think this is no longer necessary with the changes I've made. I'm gonna try the tests without the Sleep calls.

Reply:
It was possible that when you call `GetResult()`, the goroutines have not finished sending the results back.
I think this is still true for this test case, you can test it with a higher `count` and remove the `Sleep` to see if you can still get the right result.

Reply:
Increasing `count` in the tests still produced the correct result. I think because `span.Finish`, `createProfiler`, and `GetResult` are all syncing on channels, there isn't really a problem with waiting for all the various goroutines to finish anymore. I've removed the `time.Sleep` call for now, but I can put it back in if you feel it's really necessary.